### PR TITLE
GitHub: auto set inactive label

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Proposed stale action...

It labels PRs after 60 d, but never closes them.  

It does close issues after 1 year with 30 day warning.   

## PR Checklist

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by rcomer** [comment]
Obviously I approve of this sort of thing, but I thought @melissawm was working on something?

**Comment by jklymak** [comment]
I didn't see anything like that, but happy to close and let someone else put a PR in.

**Comment by oscargus** [comment]
This would close like 1200 issues. Not saying that it is a bad idea, but some of those are clearly bugs or "surprising features".

Maybe we should have some labels that does not auto-close? Confirmed bug at least? 76 would be closed https://github.com/matplotlib/matplotlib/labels/status%3A%20confirmed%20bug

Btw, are these closed as "fixed" or "not planned"?

**Comment by rcomer** [comment]
We currently have [871 open issues that have not been updated for a year](https://github.com/matplotlib/matplotlib/issues?q=is%3Aissue+is%3Aopen+updated%3A%3C2022-02-07) of which [49 were confirmed bugs](https://github.com/matplotlib/matplotlib/issues?q=is%3Aissue+is%3Aopen+updated%3A%3C2022-02-07+label%3A%22status%3A+confirmed+bug%22) so this proposal would mark those as inactive.  They would only be closed if nobody comments on them for a further month after they were labelled.

I expect that some of these bugs have actually been fixed, so I would not exempt this label.  Rather, I would expect that someone who is watching the issue will see the notification from the bot's comment and check whether a recent mpl version still has the problem.  If it does, they can comment to say so and the issue stays open for another year.

**Comment by jklymak** [comment]
I think the same for feature requests - if anyone still wants the feature, they have 30 days to comment.  We will also all see it in our notifications and can comment and perhaps reclassify or act on it as maintainers.  If no one even comments, then the request probably wasn't that important or has workarounds.

**Comment by rcomer** [comment]
For PRs, I think it would be good to find a form of words that encourages the author to at least let us know whether they plan to continue working on it.  I worry that "you can pick it up anytime" leaves it open to drift indefinitely.

**Comment by jklymak** [comment]
Note that I've also reduced the operations per run to 5. The idea being that this will lead to fewer notifications per day, and as this comes on line we may actually respond to them.  We may need to play with this depending on how many notifications we get.

File: .github/workflows/stale.yml
Comment (jklymak) [reply .github/workflows/stale.yml:9]: Can you follow up with that?  I have no idea how to do that, and I think can be completely independent of this PR.

**Comment by rcomer** [comment]
The action sifts through issues/PRs in chronological order and by default starts at the newest.  I suggest setting `ascending: true` so it starts with the oldest where it will be more likely to find things to do.

It also doesn’t remember where it got to last time it ran, so I think after the first several runs it will use up its operations just re-checking issues/PRs that it already checked.  So I think we would need to gradually increase the operations per run to help it get further.

**Comment by jklymak** [comment]
> It also doesn’t remember where it got to last time it ran, so I think after the first several runs it will use up its operations just re-checking issues/PRs that it already checked. So I think we would need to gradually increase the operations per run to help it get further.

I don't understand how it could possibly work then if it has no way of knowing what's already been done??  The default query is 30.

**Comment by jklymak** [comment]
> Sorry @jklymak for editing your comment when I was trying to reply. Is there a way to filter so that the action is only run on issues not labeled inactive.

I'm pretty sure thats how it works.

**Comment by rcomer** [comment]
If it finds the “inactive” label it then checks whether there was a comment since the label was added.  If yes, it removes the label.  If no, it checks whether the “days before close” time is up and closes if so.

**Comment by jklymak** [comment]
> Btw, are these closed as "fixed" or "not planned"?

not-planned, by default...

**Comment by rcomer** [comment]
I also don't know what counts as an "operation", but maybe it doesn't matter if it can't get through all 2000:  after the first month, it should start closing some, so then there are less than 2000 to get through.  If we also want to keep notification frequency down, we probably don't want it to find all ~900 inactive items in the first month.

File: .github/workflows/stale.yml
Comment (jklymak) [inline .github/workflows/stale.yml:22]: added exempt labels (which could be expanded) so that if we have things we don't want cleaned up we can "keep"

Comment (rcomer) [reply .github/workflows/stale.yml:22]: Maybe the "Orphaned PR" label should be exempt, as it seems redundant to check those.

Comment (jklymak) [reply .github/workflows/stale.yml]: Sure.   Not sure if I have the syntax right for the labels with spaces in them...

Comment (rcomer) [reply .github/workflows/stale.yml]: This format seems to work in Iris
https://github.com/SciTools/iris/blob/main/.github/workflows/stale.yml#L67

**Comment by jklymak** [comment]
As discussed on the call, we should probably go through with this and see if it works.  If necessary, we may need to do some of this either by hand or with offline scripts.

**Comment by rcomer** [comment]
Should the repo token be set like it is for PR welcome?
https://github.com/matplotlib/matplotlib/blob/91d8f2e882bf6fc441276c64590a54d659f2c8a4/.github/workflows/pr_welcome.yml#L16
I admit that I don’t know the difference between this and the default used by the action.
https://github.com/actions/stale#repo-token

It might also be good restrict to this repo so it doesn’t run on forks, like we do for the tests: 
https://github.com/matplotlib/matplotlib/blob/91d8f2e882bf6fc441276c64590a54d659f2c8a4/.github/workflows/tests.yml#L31

**Comment by jklymak** [comment]
Sure, did both of those.   No idea what the effect is so maybe someone with more GH Actions experience should check?

**Comment by tacaswell** [comment]
I killed appveyor for this PR to un-block running a higher-priority one for 3.7.0 prep.

**Comment by jklymak** [comment]
@rcomer, can I ping you for review?  Or we should close this if we aren't actually going to do it....

**Comment by rcomer** [comment]
I think @ksunden said he would take a look at this.  I'm far from an expert in GitHub Actions.

File: .github/workflows/stale.yml
Comment (ksunden) [inline .github/workflows/stale.yml:25]:
```suggestion
          exempt-pr-label: "keep,status: orphaned PR"
```

I didn't look further to see if they strip whitespace, but the examples given in the actions/stale README do not have a space after the comma for the comma separated labels. So just to be safe.

**Comment by ksunden** [comment]
Functionally, my only comment is the space in the comma separated label list.

Though I do want to revive @melissawm's request that these timelines appear in the docs.

File: .github/workflows/stale.yml
Comment (rcomer) [inline .github/workflows/stale.yml:8]:
```suggestion
    if: github.repository == 'matplotlib/matplotlib'
```

**Comment by jklymak** [comment]
squashed and pushed to kill CI, which doesn't do anything for this PR

**Review by ksunden — APPROVED** [review]

**Review by rcomer — APPROVED** [review]
I think this is fine to go in based on one (@ksunden) technical review - it’s not like it’s user facing code.  The messaging has had input from several people.  My only question is whether we should now leave it a couple of days before merge, so that notifications don’t pile up over the weekend.

**Comment by oscargus** [comment]
> Rather, I would expect that someone who is watching the issue will see the notification from the bot's comment and check whether a recent mpl version still has the problem.

This is the thing. I do not expect most users to check five or ten year old bug reports and, if the bug still exists, write something kind/sensible. Also, keeping track of notifications is quite hard for many people. A few days off and it is impossible to not miss something. I'm quite sure we will lose sensible reports and feature requests.

Finally, do we know if these are "closed as completed" or the other option? Ideally it should be a third option, "hide", for these, so one can check back in case one decides to work on a particular thing.

**Comment by rcomer** [comment]
> A few days off and it is impossible to not miss something. I'm quite sure we will lose sensible reports and feature requests.

I think the worst that can happen is that a currently valid issue gets closed and forgotten.  But the situation now is that we have nearly 1600 open issues, and I'm not sure it's realistic to manually sort through them all to see which are still relevant.  So either way, issues are forgotten.

If a bug or feature request affects a lot of users, then I would expect more than one user to be subscribed to the issue.  That should reduce the risk of the notification being completely missed.  Also devs can see the bot's updates and can make our own judgements if we think something should stay open.  @jklymak has deliberately chosen settings to keep the frequency low and therefore hopefully manageable.

> Finally, do we know if these are "closed as completed" or the other option? Ideally it should be a third option, "hide", for these, so one can check back in case one decides to work on a particular thing.

They are closed as "not planned", but will also retain the "inactive" label, so could be searched that way.

**Comment by rcomer** [comment]
I just created the “inactive” and “keep” labels.  Currently in white.  Please update colour and description as you see fit!

**Comment by jklymak** [comment]
Well, that worked - caught 8 old issues (no PRs) that it marked inactive.  I marked one of them "keep".  We will see if it moves on from there....

**Comment by rcomer** [comment]
I guess the oldest PR is newer than a lot of issues so it will take a long time to get to any PRs.  Maybe I shouldn’t have suggested to set `ascending: true`?

**Comment by oscargus** [comment]
I've checked the eight and am tempted to say that they are all valid requests (not sure if the colormap one is still an issue though). Not sure if I should mark the non-marked ones though...

What about adding a "closed but relevant" tag? In that way we can only keep "active" issues, but still access unsolved issues when needed. Like the mathtext issues in case we get a GSoC student working on it?

**Comment by rcomer** [comment]
Maybe in a month's time when things start to close, we could review what was closed and what we think about it.  Certainly I see no downside to adding whatever labels you think are useful to the closed issues.

**Comment by jklymak** [comment]
There is "valid" and then there is "reasonable".  If an issue is open for 12 y folks have clearly been making do without the request.  But for sure, if anyone thinks an issue will get acted on, they should comment or mark as "keep".

**Comment by rcomer** [comment]
Notes from the action’s logs on operations used:

- Newly labelling an issue uses 4. [example](https://github.com/matplotlib/matplotlib/actions/runs/4319600038/jobs/7538960502#step:2:928)
- Checking an already labelled issue uses 2. [example](https://github.com/matplotlib/matplotlib/actions/runs/4319600038/jobs/7538960502#step:2:258)
- Checking an already labelled issue and removing the label uses 3. [example](https://github.com/matplotlib/matplotlib/actions/runs/4309843539/jobs/7517681139#step:2:575)

**Pull request merged**
